index.js

```diff
@@ -8,11 +8,100 @@
 'use strict';
 
 var fs = require('fs');
+var path = require('path');
+
+/**
+ * Check if the given `filepath` exists.
+ *
+ * ```js
+ * var res = exists('package.json');
+ * console.log(res);
+ * //=> "package.json"
+ *
+ * var res = exists('fake-file.json');
+ * console.log(res)
+ * //=> false
+ * ```
+ *
+ * @name exists
+ * @param  {String} `filepath` filepath to check for.
+ * @return {String|Boolean} Returns the found filepath if it exists, otherwise returns `false`.
+ * @api public
+ */
 
 module.exports = function(filepath) {
+  if (!filepath || (typeof filepath !== 'string')) {
+    return false;
+  }
+
   try {
     (fs.accessSync || fs.statSync)(filepath);
-    return true;
+    return filepath;
   } catch (err) {}
+
+  if (process.platform === 'linux') {
+    return exists(filepath);
+  }
   return false;
 };
+
+/**
+ * Check if the filepath exists by falling back to reading in the entire directory.
+ * Returns the real filepath (for case sensitive file systems) if found.
+ *
+ * @param  {String} `filepath` filepath to check.
+ * @return {String|Boolean} Returns found filepath if exists, otherwise false.
+ */
+
+function exists(filepath) {
+  filepath = path.resolve(filepath);
+  var res = tryReaddir(filepath);
+  if (res === null) {
+    return false;
+  }
+
+  // "filepath" is a directory, an error would be
+  // thrown if it doesn't exist. if we're here, it exists
+  if (res.path === filepath) {
+    return res.path;
+  }
+
+  // "fp" is not a directory
+  var lower = filepath.toLowerCase();
+  var len = res.files.length;
+  var idx = -1;
+
+  while (++idx < len) {
+    var fp = path.resolve(res.path, res.files[idx]);
+    if (filepath === fp || lower === fp) {
```

Comment:
is it more efficient to do these equality checks before doing `fp.toLowerCase` and then checking again?